### ISSUE TITLE
User-defined, searchable, possibly public, properties (yes/no, numeric, text, date)

### ISSUE THREAD

**Issue by KayRJay** [ericberman/MyFlightbookWeb#355]
**Motivation:**

MFB has hundreds of "additional properties" for flights, many probably obvious and of general interest, and many others perhaps requested by users and some used by only a few users.  All of these are currently either numeric or yes/no.   If users could define their own properties, you might never again have to do so, and a web interface would eliminate the need to directly edit database tables.  

Some examples one or a few users might like to have even if there is not a strong general use case: pilot and co-pilot (or flight engineer) names,  instructor name, passenger names, departure and arrival time of day, take off pressure density, OAT at takeoff and landing, icing encountered in flight, use of de-icing, aircraft was rented, cost of rental, club plane, etc., etc., etc., etc. 

(I do see that some of these are already supported, buried in a long list.   We've discussed in email a rather long list of aircraft (or model) "features" that could be added by users as well as possible properties of flights, etc.).

Some user-defined properties might be of interest to only one person.   However, specific types of pilots (military, crop dusters, CFIs, test pilots, aerobatic etc.) might desire similar sets of properties uninteresting to other users.   Lack of broad interest in having MFB support a property should not prelude that support.

Permitting user-defined text properties, and allowing them to be searched, would be a robust way of supporting structured data for information that currently can only be added in comments and searched using text search.  Structured information is more likely to be remembered and entered in a consistent way.

**Basic features**

On the Add new flight page, in a dedicated group user-defined properties, there would be a pull-down menu of user-defined properties, similar to "Select a property to edit/add", along with the ability to search properties.

These user-defined properties most probably should be kept in their own group of properties, separate from those groups now on the Add flight page or Search page.    However, it is worth considering allowing user-defined properties to be created in other groups of properties like aircraft, models, categories/classes, etc.

**Searching**

User-defined properties should be searchable (ALL/ANY/NONE).  Search criteria should depend on the data type:  yes/no -> click/no click (NULL, NOT NULL); numeric -> =, <, <=. >, >=, between (when range searches in general are supported; non-zero or not until then?); text -> contains, does not contain; date (similar to numeric or like present Date search).  If user-defined properties were permitted in existing property groups, they would be entered and queried in the same sections of the Add flight and Search pages.

**Additional features**

User-defined properties could be private or public if the user wishes.  Public properties provide the benefit of crowdsourcing.    On the Add flight page, pull-down lists of properties should distinguish between public and private properties (e.g., with an asterisk for public properties).

A user-defined property could have a description that would be available with the [?] like existing "additional properties" on the Add flight page.    It would also be handy if the description appeared in pull-down lists of properties to help users decide which properties they want to add to a flight.  Descriptions of properties would be especially useful for public user-defined properties.  

A property description would be added when the property was defined, and could be edited/changed.  Descriptions of public properties could be changed by any user -- assumes people are "responsible"!

Consider creating named collections of user-defined properties: general, aerobatic, military, flight, etc. and more on request.  It's probably unnecessary/unwise to allow users to create such collections.  Chaos might ensue!   Note: collections of properties are different from the existing groups of properties users can Add to flights, and Search.

Also consider migrating some of the 600+ existing "additional properties" into one or more of these collections (even allowing duplicates across collections).   This would make it easier for a user to find an existing property he might want to add.

Export and import should support user-defined properties.  Reports could also include user-defined properties, but this would require revamping the reporting subsystem (and the compact view of the log probably suffices anyway).

**Extra comments**

You might want to record these ideas in several separate Github issues, or create new issues for selected ideas that refer to this one.  It seemed sensible to have the entire set of concepts described in one place.

Lots of ideas and much work here, but these concepts of course could be implemented and made avalable in phases.

**Comment by ericberman** [comment]
There are really three separate features proposed here:

1. User defined properties
2. Searching on property values, more than just their presence/absence
3. Grouping

I think some background would help: 
 * You indicate 2 kinds of properties (numeric and yes/no); there are actually 6 kinds of properties today: integers (i.e., counts), decimals (both times and non-times), boolean (yes/no), Date (interestingly, currently unused), Date/Datetime, Text, and Currency (like a decimal, but formatted differently)
 * I also have quite a few flags that aid in the semantics of a property.  Examples: does this property count as a flight review (checkrides typically do), should it automatically be excluded from the list of previously used values (you only ever take one CFII checkride), if it's numeric should it be totaled (e.g., hours of solo time should show up in totals, but it's nonsensical to add your tach time together), whether a decimal value is a time (in which case it should respect your preference whether to display as hours:minutes vs. decimal) or a simple decimal (quarts of oil added is always displayed in decimal) and so forth.
 * Adding a new property involves no direct database entry: I use a web interface to set the title, the format string (how it is displayed in-situ, which is different from how it is displayed for editing.  E.g., Solo Time is "Solo Time:" as the prompt for editing, but it is "Solo Time: 1.3" in-situ (with a value).  I add new properties fairly frequently, and it is a trivial and safe operation.  The hard part is not technical; it is data curation: ensuring that the new property is not a duplicate of an existing property, and that it has the right semantics (e.g., do you need to indicate "I performed chandelles on this flight", which is a yes/no value, or do you need to record the number of chandelles you performed, which is an integer?)
 * And to your question about "Edit" in the "Add/Edit" prompt: you're adding a property to the form, presumably so that you can then change (edit) its value.  I could change it to simply "add", I suppose, since the "edit" is kinda obvious.  I'll do that.

So, for the feature requests:
*User Defined Properties*
While I've had requests for this numerous times in the past, people have always been perfectly happy with me creating the properties they need.  It's a form of crowd-sourcing where one pilot's request makes the property available for other pilots, but I do the curation to ensure that the semantics and such above are followed.  I've trailed off to adding a new property perhaps once a month or so, which suggests that I've got pretty good coverage.

Once created, I *never* allow a property to be edited in a substantial way (fixing typos, editing description, etc. are not substantial; changing renaming "solo time" to "night vision time" would be a nightmare).  People occasionally actually do these sorts of crazy edits with models or aircraft (re-using the Cirrus SR22 model and changing the fields to be a Piper PA28, for example), which is fortunately super easy to undo; but if this happened in properties and someone used that in their flights, then it's essentially impossible to fix.  So I'm super leery of enabling the bypass of the curation.  Yes, that's possible with models and aircraft (and I spend a lot of time curating that), but pilots know their models and aircraft and the semantics around those are less esoteric than around properties. E.g., asking for the name of the model or the tail number is something simple for pilots to answer; asking them whether a given property is decimal vs. integer is more abstract (and then there are the 20 or so flags above I'd need to explain to them).  I can edit a model or aircraft after a user edits it and fix things up; but if someone creates a problematic property, I can't do much because people may have used those properties in their flights by the time I get a chance to review the change, and that's a really intractable problem to fix up.

Enabling privately defined properties would be more doable and I have considered a feature where I reserve some properties as being user-definable, private to that user.  The two main reasons I haven't (besides the work) comes back to the semantics/flags issues I describe above, but also to the fact that if a property is useful for one pilot, it's probably useful to others, so why don't I just add the property and make it available to everyone?  The set of properties is not nearly as unbounded as the set of models or aircraft, and the curation needs are much higher; I'm happy to play that curation role.

So for those reasons, I'm inclined to remain "won't fix" on this feature.

*Searching on values*
This is issue #326 , straight up.  It's a reasonable feature, but there's nothing in this github issue that isn't in that one.

*Grouping*
We've also discussed this before, and the Templates feature covers a lot of it: your suggestion of creating named-collections of properties is precisely what the template feature enables.  (And there's no chaos from having it be user-defined; I have that already and it works fine.  Plus if someone creates a garbage grouping, it's just a garbage grouping - it doesn't impact anybody else unless that other person chooses to use it, and then they'll see it's garbage and stop using it.)

You suggest migrating the properties into these collections; Templates can provide precisely that functionality as well, and a given property can be in multiple templates.  

I suppose I could do a naming/grouping taxonomy that is distinct from templates, more for the purpose of finding a property than for the purpose of selecting a group of them at a time, but I'm not sure that isn't overkill, especially since you can already do a simple type-ahead search to quickly find a property that you're looking for.  After all, if you're looking for "Night Vision - Goggle Operations", it's nice that it's under a "Military" grouping or a "Night flight" grouping, but those ultimately are just mechanisms to find it; if typing "gogg" finds it, then that solves the problem too.

So, for now at least, I think the combination of templates and type-ahead search addresses this.

**Comment by KayRJay** [comment]
Thanks for the detailed and explanation!
